torch/lib/THD/master_worker/common/CommandChannel.cpp

```diff
@@ -80,23 +88,91 @@ bool MasterCommandChannel::init() {
    // close listen socket
   ::close(_sockets[0]);
   _sockets[0] = -1;
+
+  _started = true;
+  _error_thread = std::thread(&MasterCommandChannel::errorHandler, this);
   return true;
 }
 
+void MasterCommandChannel::errorHandler() {
+  while (true) {
+    auto error = recvError();
+    if (_exiting) {
+      return;
+    }
+
+    _error.reset(new std::string(
+      "error (rank " + std::to_string(std::get<0>(error)) + "): " + std::get<1>(error)
+    ));
+  }
+}
+
 void MasterCommandChannel::sendMessage(std::unique_ptr<rpc::RPCMessage> msg, int rank) {
+  // Throw error received from a worker.
+  if (_error) {
+    throw std::runtime_error(*_error);
+  }
+
   if ((rank <= 0) || (rank >= _sockets.size())) {
     throw std::domain_error("sendMessage received invalid rank as parameter");
   }
 
   ::thd::sendMessage(_sockets[rank], std::move(msg));
 }
 
-std::unique_ptr<rpc::RPCMessage> MasterCommandChannel::recvMessage(int rank) {
-  if ((rank <= 0) || (rank >= _sockets.size())) {
-    throw std::domain_error("recvMessage received invalid rank as parameter");
+std::tuple<rank_type, std::string> MasterCommandChannel::recvError() {
+  if (!_poll_events) {
+    // cache poll events array, it will be reused in another `receiveError` calls
+    _poll_events.reset(new struct pollfd[_sockets.size()]);
+    for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+      _poll_events[rank] = {
+        .fd = _sockets[rank],
+        .events = POLLIN
+      };
+    }
+  }
+
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    _poll_events[rank].revents = 0;
+  }
+
+  try {
+    int ret = 0;
+    while (ret == 0) { // loop until there is data to read
+      SYSCHECK(ret = ::poll(_poll_events.get(), _sockets.size(), 500))
+
+      if (_exiting) {
+        return std::make_tuple(0, "");
+      }
+    }
+  } catch (const std::exception& e) {
+    return std::make_tuple(0, "poll: " + std::string(e.what()));
   }
 
-  return receiveMessage(_sockets[rank]);
+  for (std::size_t rank = 0; rank < _sockets.size(); ++rank) {
+    if (this->_poll_events[rank].revents == 0)
+      continue;
+
+    if (_poll_events[rank].revents ^ POLLIN) {
+      _poll_events[rank].fd = -1; // mark worker as ignored
+      return std::make_tuple(rank, "connection with worker has been closed");
+    }
+
+    try {
+      // receive error
+      std::uint64_t error_length;
+      recv_bytes<std::uint64_t>(_poll_events[rank].fd, &error_length, 1);
+
+      std::unique_ptr<char[]> error(new char[error_length + 1]);
```

Comment:
Ahh, sorry that `+ 1` is totally unnecessary and misleading. Fortunately, string (line below) is created by taking predefined `error_length` so it is not relying on null termination.